pkg/metricsadapter/provider/custommetrics.go

```diff
@@ -3,33 +3,285 @@ package provider
-	return nil, fmt.Errorf("karmada-metrics-adapter still not implement it")
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return nil, err
+	}
+	metricValueList := &custom_metrics.MetricValueList{}
+	for _, cluster := range clusters {
+		metrics, err := c.getMetricByName(ctx, cluster.Name, name, info, metricSelector)
+		if err != nil {
+			klog.Warningf("query %s's %s metric from cluster %s failed, err: %+v", info.GroupResource.String(), info.Metric, cluster.Name, err)
+			continue
+		}
+		metricValueList.Items = append(metricValueList.Items, metrics.Items...)
+	}
+	if len(metricValueList.Items) != 1 {
+		return nil, fmt.Errorf("the custom metrics API server returned %v results when we asked for exactly one", len(metricValueList.Items))
+	}
+	return &metricValueList.Items[0], nil
 }
 
 // GetMetricBySelector will query metrics by selector from member clusters and return the result
 func (c *CustomMetricsProvider) GetMetricBySelector(ctx context.Context, namespace string, selector labels.Selector, info provider.CustomMetricInfo, metricSelector labels.Selector) (*custom_metrics.MetricValueList, error) {
-	return nil, fmt.Errorf("karmada-metrics-adapter still not implement it")
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return nil, err
+	}
+	metricValueList := &custom_metrics.MetricValueList{}
+	for _, cluster := range clusters {
+		metrics, err := c.getMetricBySelector(ctx, cluster.Name, namespace, selector, info, metricSelector)
+		if err != nil {
+			klog.Warningf("query %s's %s metric from cluster %s failed", info.GroupResource.String(), info.Metric, cluster.Name)
+			continue
+		}
+		metricValueList.Items = append(metricValueList.Items, metrics.Items...)
+	}
+	return metricValueList, nil
+}
+
+func (c *CustomMetricsProvider) getPreferredVersion(discoveryClient *discovery.DiscoveryClient) (schema.GroupVersion, error) {
+	apiGroups, err := discoveryClient.ServerGroups()
+	if err != nil {
+		return schema.GroupVersion{}, err
+	}
+	if gv, support := supportedMetricsAPIVersionAvailable(apiGroups); support {
+		return gv, nil
+	}
+	return schema.GroupVersion{}, fmt.Errorf("custom.metrics.k8s.io not found")
+}
+
+func (c *CustomMetricsProvider) getMetricByName(ctx context.Context, clusterName string, name types.NamespacedName, info provider.CustomMetricInfo, metricSelector labels.Selector) (*custom_metrics.MetricValueList, error) {
+	// handle namespace separately
+	if info.GroupResource.Resource == "namespaces" && info.GroupResource.Group == "" {
+		return c.getForNamespace(ctx, clusterName, name.Name, info.Metric, metricSelector)
+	}
+	discoveryClient, err := c.newDiscoveryClientFunc(clusterName)
```

Comment:
I believe we can obtain the necessary information from the cluster API.  In order to expedite the process

Reply:
Update.